Csrf fix [WIP]

ieaves opened the pull request:
Changes:
* Substantial improvements to server configuration supporting much richer user customization over the running environment
* Fixes CORS for applications deployed outside of localhost
* Fixes CSRF for application deployed out of localhost
* Sensible default configurations for deployment environments when `SERVER_HOST` and `FRONTEND_PORT` configurations are provided
* Dynamic configuration on the frontend to specify server deployment location
* example deployment configs for Kubernetes

edlouth commented on grai-server/app/the_guide/settings.py:
```diff
@@ -14,21 +15,55 @@
+    default_cors_origins = [
+        f"{scheme}://{host}:{port}"
+        for scheme, host, port in product(schemes, default_hosts, default_ports)
+    ]
```
Is this crossing all three of these arrays?  I guess it works

ieaves replied:
Yes, exactly. It's producing every combination of scheme, host, and port.

edlouth commented on .gitignore:
```diff
@@ -5,7 +5,12 @@
+=======
+.fleet/
+**/.fleet/
+>>>>>>> Stashed changes
```
Merge error to fix

ieaves replied:
bah, good catch. thank you.

edlouth commented on grai-server/app/the_guide/settings.py:
```diff
@@ -88,7 +123,7 @@ def clean_allowed_hosts(val):
     "django.contrib.sessions.middleware.SessionMiddleware",
     "django.middleware.common.CommonMiddleware",
-    "django.middleware.csrf.CsrfViewMiddleware",
+    # "django.middleware.csrf.CsrfViewMiddleware",
```
Is this just removing CSRF?

ieaves replied:
I'm not sure if it completely disables CSRF but the [docs](https://docs.djangoproject.com/en/4.1/ref/csrf/) describe it providing a specific form of subdomain protection and I wanted to see if that was responsible for the issue I was seeing when deployed to the server.

edlouth commented:
> Oh fuck yes! That solves the CORS issue on the frontend. I need to figure out why the admin is still generating a CSRF verification issue logging into the admin before we can merge though.
> 
> ![Screenshot 2022-11-21 at 11 23 53 PM](https://user-images.githubusercontent.com/7785207/203228847-f7a85c10-5ba9-48c4-85d0-660e18eb7715.png)

I think this can be solved by setting `SERVER_HOST=grai.tinglacier.com`

ieaves commented:
That was annoyingly the configuration I was using in that commit. It's sorted now but the issue ended up being that I only included `scheme://host:port` combinations into `CORS_ALLOWED_ORIGINS` when I needed a record omitting the port suffix. 

There were a few logic bugs I've fixed and cleaned up the CORS_ALLOWED_ORIGINS and CSRF_TRUSTED_ORIGINS so that they aren't maximally permissive.
